core/eth/state.go

```diff
@@ -1,82 +1,309 @@
+		return nil, fmt.Errorf("transaction %s is still pending for operator socket update event", txHash.Hex())
+	}
+	return transaction, nil
+}
+
+func (cs *ChainState) parseOperatorSocketUpdate(callData []byte, log *types.Log) (core.OperatorID, string, error) {
+	operatorID, err := cs.parseOperatorIDFromLog(log)
+	if err != nil {
+		cs.logger.Warn("failed to parse operator ID from log. skipping malformed log",
+			"error", err)
+	}
+
+	socket, err := cs.parseSocketFromCallData(callData)
+	if err != nil {
+		cs.logger.Warn("failed to parse socket update event. skipping malformed log",
+			"error", err,
+			"txHash", log.TxHash.Hex(),
+			"operatorId", operatorID)
+	}
+	return operatorID, socket, nil
+}
+
+// parseOperatorIDFromLog parses the operator ID from a log and returns the operator ID.
+func (cs *ChainState) parseOperatorIDFromLog(log *types.Log) (core.OperatorID, error) {
+	if len(log.Topics) < 2 {
+		return core.OperatorID{}, fmt.Errorf("log topics too short: expected at least 2 topics, got %d", len(log.Topics))
+	}
+	if len(log.Topics[1].Bytes()) != 32 {
+		return core.OperatorID{}, fmt.Errorf("operatorID is expecting 32 bytes, got %d", len(log.Topics[1].Bytes()))
+	}
+	operatorID := core.OperatorID(log.Topics[1].Bytes())
+	return operatorID, nil
+}
+
+// parseSocketFromCallData parses the socket update event from a log and returns the operator ID and socket address.
+func (cs *ChainState) parseSocketFromCallData(calldata []byte) (string, error) {
+	// Add length check for method name and input data
+	if len(calldata) <= 4 {
+		return "", fmt.Errorf("calldata too short: expected more than 4 bytes for method name and input data, got %d bytes", len(calldata))
+	}
+
+	rcAbi, err := abi.JSON(bytes.NewReader(common.RegistryCoordinatorAbi))
+	if err != nil {
+		return "", err
+	}
+	methodSig := calldata[:4]
+	method, err := rcAbi.MethodById(methodSig)
+	if err != nil {
+		return "", err
+	}
+
+	inputs, err := method.Inputs.Unpack(calldata[4:])
 	if err != nil {
 		return "", err
 	}
+
+	var socket string
+	if (method.Name == "registerOperator" || method.Name == "registerOperatorWithChurn") && len(inputs) >= 2 {
+		socket = inputs[1].(string)
+	} else if method.Name == "updateSocket" && len(inputs) >= 1 {
+		socket = inputs[0].(string)
+	} else {
+		// this should never happen; we are going to return empty string
+		cs.logger.Warn("method and input length mismatch for socket update event: %s", "method", method.Name)
+		return "", fmt.Errorf("method and input length mismatch for socket update event: %s", method.Name)
+	}
 	return socket, nil
 }
 
-func getOperatorState(operatorsByQuorum core.OperatorStakes, blockNumber uint32) (*core.OperatorState, error) {
+// refreshSocketMap refresh the socket map for the given operators by quorums at the current block.
+func (cs *ChainState) refreshSocketMap(creader context.Context, operatorsByQuorum core.OperatorStakes) error {
```

Comment:
There are still lots of instances where the context is named `creader`, instead of `ctx`